plugins/genesys/fu-genesys-usbhub-device.c

```diff
@@ -609,11 +611,61 @@ fu_genesys_usbhub_device_get_fw_size(FuGenesysUsbhubDevice *self, int bank_num,
 		g_prefix_error(error, "error getting fw size from device: ");
 		return FALSE;
 	}
-	self->code_size = 1024 * kbs;
+	self->code_size[bank_num] = 1024 * kbs;
 
 	/* success */
 	return TRUE;
 }
+
+static gboolean
+fu_genesys_usbhub_device_get_fw_version(FuDevice *device, int bank_num, GError **error)
+{
+	FuGenesysUsbhubDevice *self = FU_GENESYS_USBHUB_DEVICE(device);
+	gsize address = self->fw_bank_addr[bank_num];
+	gsize bufsz;
+	guint16 version_raw;
+	g_autoptr(FuFirmware) firmware = NULL;
+	g_autofree guint8 *buf = NULL;
+
+	g_return_val_if_fail(bank_num < 2, FALSE);
+
+	if (self->code_size[bank_num] == 0)
+		if (!fu_genesys_usbhub_device_get_fw_size(self, bank_num, error))
+			return FALSE;
+
+	if (!fu_genesys_usbhub_device_check_fw_signature(self, bank_num, error))
```

Comment:
I found `fu_genesys_usbhub_device_check_fw_signature()` already checked in `fu_genesys_usbhub_device_get_fw_size()`, can I add `else{}` to separate them?

Reply:
Well, I do not like that new function `fu_genesys_usbhub_device_get_fw_version()`.

I prefer you copy/paste `fu_genesys_usbhub_device_get_fw_size()`, create and use the new `GENESYS_USBHUB_VERSION_OFFSET` instead.

``` c
/* read the firmware vresion from the firmware stored in the device */
static gboolean
fu_genesys_usbhub_device_get_fw_version(FuGenesysUsbhubDevice *self, int bank_num, GError **error)
{
	guint16 version_raw = 0;

	g_return_val_if_fail(bank_num < 2, FALSE);
	g_return_val_if_fail(self->fw_bank_vers[bank_num] == 0, FALSE);

	if (!fu_genesys_usbhub_device_check_fw_signature(self, bank_num, error))
		return FALSE;

	/* get firmware version from device */
	if (!fu_genesys_usbhub_device_read_flash(self,
						 self->fw_bank_addr[bank_num] +
						     GENESYS_USBHUB_VERSION_OFFSET,
						 &kbs,
						 1,
						 NULL,
						 error)) {
		g_prefix_error(error, "error getting fw version from device: ");
		return FALSE;
	}
	if (version_raw != 0xffff)
		self->fw_bank_vers[bank_num] = version_raw;

	/* success */
	return TRUE;
}
```

But the whole thing is undoing the verifying of firmware integrity. Is that correct to not verify the firmware integrity?

Reply:
> But the whole thing is undoing the verifying of firmware integrity. Is that correct to not verify the firmware integrity?

This function's goal is to get the firmware version of the valid firmware bank. So verify the firmware integrity before get the version.

The verify flow should likes:
1. check signature(XROM for hub).
2. get firmware code size if it is not fixed.
3. read entire firmware.
4. calculate and match checksum.
--- 
Therefore,  I think we can modify below.
1. remove calling `fu_genesys_usbhub_device_check_fw_signature()` in fu_genesys_usbhub_device_get_fw_size().
2. modify in `fu_genesys_usbhub_device_get_fw_version()`
```
	if (!fu_genesys_usbhub_device_check_fw_signature(self, bank_num, error))
		return FALSE;

	if (self->code_size[bank_num] == 0) {
		if (!fu_genesys_usbhub_device_get_fw_size(self, bank_num, error))
			return FALSE;
	}
```
3. rename `fu_genesys_usbhub_device_get_fw_version()`, I think it is confuse for other. Which one is better, get_bank_version or query_bank_fw_version? Do you have any idea?

Reply:
What was wrong with the previous code already? the one in the `_setup()` vfunc?
  
> 1. check signature(XROM for hub).
> 1. get firmware code size if it is not fixed.
> 1. read entire firmware.
> 1. calculate and match checksum.

I guess the `_setup()` should handle the check point 2.

Can we query the running bank to the firmware? If so, we can get rid of points 3. and 4.

Reply:
> Let's step back a bit; what was wrong with the previous code already?

The fu_genesys_usbhub_device_check_fw_signature() already called in fu_genesys_usbhub_device_get_fw_size(), it may check twice when `if (self->code_size[bank_num] == 0)`. 
And it becomes wrong if I remove calling fu_genesys_usbhub_device_check_fw_signature() in fu_genesys_usbhub_device_get_fw_size(). Therefore I have to change the calling serial.

> Can we query the running bank to the firmware? If so, we can get rid of points 3. and 4.

Yes, we already set into `_FuGenesysUsbhubDevice::running_bank`. We can check the not running bank only.

Reply:
> The fu_genesys_usbhub_device_check_fw_signature() already called in fu_genesys_usbhub_device_get_fw_size(), it may check twice when if (self->code_size[bank_num] == 0).
And it becomes wrong if I remove calling fu_genesys_usbhub_device_check_fw_signature() in fu_genesys_usbhub_device_get_fw_size(). Therefore I have to change the calling serial.

... I was thinking even further. What was wrong in the code that was in the `_setup()` vfunc (before your changes)? because, if you do not need those current changes, we can leave it as it was, merge the PR and clean that code afterward.

> Yes, we already set into _FuGenesysUsbhubDevice::running_bank. We can check the not running bank only.

Nice. This would be nice to update the `_dump_firmware()` to dump the running bank instead of the first one; we can do it later.

Reply:
Perfect. We will do the change after the PR is merged.